From cf46b1017a2663c59f07abd417573b9580f0145f Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Tue, 29 Jul 2014 09:14:15 -0400
Subject: [PATCH] dri: Stop (uselessly) initializing the ValidateTree hooks

Signed-off-by: Adam Jackson <ajax@redhat.com>
---
 src/mga_dri.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/src/mga_dri.c b/src/mga_dri.c
index a1c39f3..b957028 100644
--- a/src/mga_dri.c
+++ b/src/mga_dri.c
@@ -1165,9 +1165,6 @@ Bool MGADRIScreenInit( ScreenPtr pScreen )
    pDRIInfo->frameBufferStride = pScrn->displayWidth*(pScrn->bitsPerPixel/8);
    pDRIInfo->ddxDrawableTableEntry = MGA_MAX_DRAWABLES;
 
-   pDRIInfo->wrap.ValidateTree = NULL;
-   pDRIInfo->wrap.PostValidateTree = NULL;
-
    pDRIInfo->createDummyCtx = TRUE;
    pDRIInfo->createDummyCtxPriv = FALSE;
 
-- 
2.5.0

