diff -up xf86-video-mga-1.6.3/src/mga_driver.c.jx xf86-video-mga-1.6.3/src/mga_driver.c
--- xf86-video-mga-1.6.3/src/mga_driver.c.jx	2014-08-07 11:46:19.000000000 -0400
+++ xf86-video-mga-1.6.3/src/mga_driver.c	2014-08-07 11:48:02.472566007 -0400
@@ -1786,7 +1786,9 @@ MGAPreInit(ScrnInfoPtr pScrn, int flags)
 	pScrn->videoRam = MGACountRam(pScrn);
     }
 
-    if (pMga->is_G200SE && pScrn->videoRam < 2048)
+    if ((pMga->is_G200SE && pScrn->videoRam < 2048 ||
+	 pMga->Chipset == PCI_CHIP_MGAG200_SE_A_PCI) &&
+        pScrn->confScreen->defaultdepth == 0)
 	pScrn->confScreen->defaultdepth = 16;
 
     if (!xf86SetDepthBpp(pScrn, 0, 0, 0, flags24)) {
