--- xf86-video-mga-1.6.3/src/mga_driver.c.g200eh3	2016-12-14 04:59:57.118997677 -0500
+++ xf86-video-mga-1.6.3/src/mga_driver.c	2016-12-14 05:24:15.501997550 -0500
@@ -425,6 +425,21 @@ static const struct mga_device_attribute
 	},
 
 	16384, 0x4000,          /* Memory probe size & offset values */
+    },
+
+    [17] = { 0, 1, 0, 0, 1, 0, 0, new_BARs,
+            (TRANSC_SOLID_FILL | TWO_PASS_COLOR_EXPAND | USE_LINEAR_EXPANSION),
+	{
+	    { 50000, 230000 }, /* System VCO frequencies */
+	    { 50000, 203400 }, /* Pixel VCO frequencies */
+	    { 0, 0 },          /* Video VCO frequencies */
+	    45000,            /* Memory clock */
+	    27050,             /* PLL reference frequency */
+	    0,                 /* Supports fast bitblt? */
+	    MGA_HOST_PCI       /* Host interface */
+	},
+
+	16384, 0x4000,          /* Memory probe size & offset values */
     }
 };
 
@@ -458,6 +472,8 @@ static const struct pci_id_match mga_dev
 
     MGA_DEVICE_MATCH( PCI_CHIP_MGAG200_EW3_PCI, 16 ),
 
+    MGA_DEVICE_MATCH( PCI_CHIP_MGAG200_EH3_PCI, 17 ),
+
     { 0, 0, 0 },
 };
 #endif
@@ -479,6 +496,7 @@ static SymTabRec MGAChipsets[] = {
     { PCI_CHIP_MGAG200_WINBOND_PCI,	"mgag200 eW Nuvoton" },
     { PCI_CHIP_MGAG200_EW3_PCI,	"mgag200 eW3 Nuvoton" },
     { PCI_CHIP_MGAG200_EH_PCI,	"mgag200eH" },
+     { PCI_CHIP_MGAG200_EH3_PCI,	"mgag200eH3" },
     { PCI_CHIP_MGAG400,		"mgag400" },
     { PCI_CHIP_MGAG550,		"mgag550" },
     {-1,			NULL }
@@ -507,6 +523,8 @@ static PciChipsets MGAPciChipsets[] = {
 	RES_SHARED_VGA },
     { PCI_CHIP_MGAG200_EW3_PCI, PCI_CHIP_MGAG200_EW3_PCI,
 	RES_SHARED_VGA },
+    { PCI_CHIP_MGAG200_EH3_PCI, PCI_CHIP_MGAG200_EH3_PCI,
+	RES_SHARED_VGA },
     { PCI_CHIP_MGAG400,	    PCI_CHIP_MGAG400,	RES_SHARED_VGA },
     { PCI_CHIP_MGAG550,	    PCI_CHIP_MGAG550,	RES_SHARED_VGA },
     { -1,			-1,		RES_UNDEFINED }
@@ -694,6 +712,7 @@ MGAPciProbe(DriverPtr drv, int entity_nu
 	    case PCI_CHIP_MGAG200_WINBOND_PCI:
 	    case PCI_CHIP_MGAG200_EW3_PCI:
             case PCI_CHIP_MGAG200_EH_PCI:
+            case PCI_CHIP_MGAG200_EH3_PCI:
 		xf86DrvMsg(0, X_ERROR,
 	                   "mga: The PCI device 0x%x at %2.2d@%2.2d:%2.2d:%1.1d has a kernel module claiming it.\n",
 	                   dev->device_id, dev->bus, dev->domain, dev->dev, dev->func);
@@ -948,6 +967,10 @@ MGAProbe(DriverPtr drv, int flags)
                 attrib_no = 16;
                 break;
 
+            case PCI_CHIP_MGAG200_EH3_PCI:
+                attrib_no = 17;
+                break;
+
 	    default:
 		return FALSE;
             }
@@ -1577,7 +1600,8 @@ MGAPreInit(ScrnInfoPtr pScrn, int flags)
     pMga->is_G200EV = (pMga->Chipset == PCI_CHIP_MGAG200_EV_PCI);
     pMga->is_G200WB = (pMga->Chipset == PCI_CHIP_MGAG200_WINBOND_PCI)
 	|| (pMga->Chipset ==  PCI_CHIP_MGAG200_EW3_PCI);
-    pMga->is_G200EH = (pMga->Chipset == PCI_CHIP_MGAG200_EH_PCI);
+    pMga->is_G200EH = (pMga->Chipset == PCI_CHIP_MGAG200_EH_PCI)
+        ||  (pMga->Chipset ==  PCI_CHIP_MGAG200_EH3_PCI);
     pMga->is_G200ER = (pMga->Chipset == PCI_CHIP_MGAG200_ER_PCI);
 
     pMga->DualHeadEnabled = FALSE;
@@ -1605,6 +1629,7 @@ MGAPreInit(ScrnInfoPtr pScrn, int flags)
     case PCI_CHIP_MGAG200_WINBOND_PCI:
     case PCI_CHIP_MGAG200_EW3_PCI:
     case PCI_CHIP_MGAG200_EH_PCI:
+    case PCI_CHIP_MGAG200_EH3_PCI:
 	pMga->HWCursor = FALSE;
 	xf86DrvMsg(pScrn->scrnIndex, X_INFO,
 		   "HW cursor is not supported with video redirection on"
@@ -2026,6 +2051,7 @@ MGAPreInit(ScrnInfoPtr pScrn, int flags)
         case PCI_CHIP_MGAG200_EW3_PCI:
 	case PCI_CHIP_MGAG200_EV_PCI:
 	case PCI_CHIP_MGAG200_EH_PCI:
+	case PCI_CHIP_MGAG200_EH3_PCI:
 	case PCI_CHIP_MGAG200_ER_PCI:
 	    pMga->randr12 = TRUE;
 	    break;
@@ -2097,6 +2123,7 @@ MGAPreInit(ScrnInfoPtr pScrn, int flags)
     case PCI_CHIP_MGAG200_EV_PCI:
     case PCI_CHIP_MGAG200_EH_PCI:
     case PCI_CHIP_MGAG200_ER_PCI:	
+    case PCI_CHIP_MGAG200_EH3_PCI:
     case PCI_CHIP_MGAG400:
     case PCI_CHIP_MGAG550:
 	MGAGSetupFuncs(pScrn);
@@ -2206,7 +2233,8 @@ MGAPreInit(ScrnInfoPtr pScrn, int flags)
           case PCI_CHIP_MGAG200_WINBOND_PCI:
           case PCI_CHIP_MGAG200_EW3_PCI:
 	  case PCI_CHIP_MGAG200_EV_PCI:
-      case PCI_CHIP_MGAG200_EH_PCI:
+          case PCI_CHIP_MGAG200_EH_PCI:
+          case PCI_CHIP_MGAG200_EH3_PCI:
 	  case PCI_CHIP_MGAG200_ER_PCI:	  
 	    pMga->SrcOrg = 0;
 	    pMga->DstOrg = 0;
@@ -2419,7 +2447,8 @@ MGAPreInit(ScrnInfoPtr pScrn, int flags)
         case PCI_CHIP_MGAG200_WINBOND_PCI:
         case PCI_CHIP_MGAG200_EW3_PCI:
 	case PCI_CHIP_MGAG200_EV_PCI:
-    case PCI_CHIP_MGAG200_EH_PCI:
+        case PCI_CHIP_MGAG200_EH_PCI:
+        case PCI_CHIP_MGAG200_EH3_PCI:
 	case PCI_CHIP_MGAG200_ER_PCI:	
 	case PCI_CHIP_MGAG400:
 	case PCI_CHIP_MGAG550:
@@ -4032,7 +4061,7 @@ mga_valid_mode(ScrnInfoPtr pScrn, Displa
     } else if (pMga->is_G200EV
 	       && (xf86ModeBandwidth(mode, pScrn->bitsPerPixel) > 327)) {
 	return MODE_BANDWIDTH;
-    } else if (pMga->is_G200EH
+    } else if (pMga->is_G200EH && (pMga->Chipset != PCI_CHIP_MGAG200_EH3_PCI)
                && (xf86ModeBandwidth(mode, pScrn->bitsPerPixel) > 375)) {
         return MODE_BANDWIDTH;
     } else if (pMga->is_G200ER
--- xf86-video-mga-1.6.3/src/mga_dacG.c.g200eh3	2016-12-14 04:59:17.598997927 -0500
+++ xf86-video-mga-1.6.3/src/mga_dacG.c	2016-12-14 06:26:25.570996753 -0500
@@ -383,6 +383,52 @@ MGAG200EHComputePLLParam(long lFo, int *
     }
 }
 
+void
+MGAG200EH3ComputePLLParam(long lFo, int *M, int *N, int *P)
+{
+    unsigned int ulComputedFo;
+    unsigned int ulFDelta;
+    unsigned int ulFPermitedDelta;
+    unsigned int ulFTmpDelta;
+    unsigned int ulTestP;
+    unsigned int ulTestM;
+    unsigned int ulTestN;
+    unsigned int ulVCOMax;
+    unsigned int ulVCOMin;
+    unsigned int ulPLLFreqRef;
+
+    ulVCOMax        = 3000000;
+    ulVCOMin        = 1500000;
+    ulPLLFreqRef    = 25000;
+
+    ulTestP         = 0;
+
+    ulFDelta = 0xFFFFFFFF;
+    /* Permited delta is 0.5% as VESA Specification */
+    ulFPermitedDelta = lFo * 5 / 1000;  
+
+    /* Then we need to minimize the M while staying within 0.5% */
+    for (ulTestM = 150; ulTestM >= 6; ulTestM--) {
+        if ((lFo * ulTestM) > ulVCOMax) continue;
+        if ((lFo * ulTestM) < ulVCOMin) continue;
+
+        for (ulTestN = 120; ulTestN >= 60; ulTestN--) {
+            ulComputedFo = (ulPLLFreqRef * ulTestN) / ulTestM;
+            if (ulComputedFo > lFo)
+	        ulFTmpDelta = ulComputedFo - lFo;
+            else
+                ulFTmpDelta = lFo - ulComputedFo;
+
+            if (ulFTmpDelta < ulFDelta) {
+                ulFDelta = ulFTmpDelta;
+                *M = (CARD8)(ulTestM);
+                *N = (CARD8)(ulTestN);
+                *P = (CARD8)(ulTestP);
+            }
+        }
+    }
+}
+
 static void
 MGAG200EVPIXPLLSET(ScrnInfoPtr pScrn, MGARegPtr mgaReg)
 {
@@ -1047,7 +1093,15 @@ MGAGSetPCLK( ScrnInfoPtr pScrn, long f_o
 	    pReg->PllN = n;
 	    pReg->PllP = p;
 	} else if (pMga->is_G200EH) {
-	    MGAG200EHComputePLLParam(f_out, &m, &n, &p);
+            if (pMga->Chipset == PCI_CHIP_MGAG200_EH3_PCI)
+            {
+                 MGAG200EH3ComputePLLParam(f_out, &m, &n, &p);
+            }
+            else
+            {
+	         MGAG200EHComputePLLParam(f_out, &m, &n, &p);
+            }
+
 
 	    pReg->PllM = m;
 	    pReg->PllN = n;
@@ -1254,6 +1308,7 @@ MGAGInit(ScrnInfoPtr pScrn, DisplayModeP
 			break;
 
         case PCI_CHIP_MGAG200_EH_PCI:
+        case PCI_CHIP_MGAG200_EH3_PCI:
                 pReg->DacRegs[MGA1064_MISC_CTL] =
                     MGA1064_MISC_CTL_VGA8 |
                     MGA1064_MISC_CTL_DAC_RAM_CS;
--- xf86-video-mga-1.6.3/src/mga_merge.c.g200eh3	2016-12-14 05:01:09.254997805 -0500
+++ xf86-video-mga-1.6.3/src/mga_merge.c	2016-12-14 05:04:34.276997779 -0500
@@ -355,6 +355,7 @@ MGAPreInitMergedFB(ScrnInfoPtr pScrn1, i
     case PCI_CHIP_MGAG200_EV_PCI:
     case PCI_CHIP_MGAG200_EH_PCI:
     case PCI_CHIP_MGAG200_ER_PCI:
+    case PCI_CHIP_MGAG200_EH3_PCI:
     case PCI_CHIP_MGAG400:
     case PCI_CHIP_MGAG550:
 	MGAGSetupFuncs(pScrn);
@@ -505,6 +506,7 @@ MGAPreInitMergedFB(ScrnInfoPtr pScrn1, i
         case PCI_CHIP_MGAG200_EV_PCI:
         case PCI_CHIP_MGAG200_EH_PCI:
 	case PCI_CHIP_MGAG200_ER_PCI:		
+	case PCI_CHIP_MGAG200_EH3_PCI:		
 	case PCI_CHIP_MGAG400:
 	case PCI_CHIP_MGAG550:
 	   maxPitch = 4096;
--- xf86-video-mga-1.6.3/src/mga_randr.h.g200eh3	2016-12-14 05:01:27.166997820 -0500
+++ xf86-video-mga-1.6.3/src/mga_randr.h	2016-12-14 05:03:27.369997807 -0500
@@ -44,4 +44,5 @@ extern void MGAG200EW3ComputePLLParam(lo
 extern void MGAG200EHComputePLLParam(long lFo, int *M, int *N, int *P);
 extern void MGAG200ERComputePLLParam(long lFo, int *M, int *N, int *P);
 extern void MGAG200E4ComputePLLParam(long lFo, int *M, int *N, int *P);
+extern void MGAG200EH3ComputePLLParam(long lFo, int *M, int *N, int *P);
 #endif
--- xf86-video-mga-1.6.3/src/mga_storm.c.g200eh3	2016-12-14 04:58:33.560997851 -0500
+++ xf86-video-mga-1.6.3/src/mga_storm.c	2016-12-14 05:02:50.228997796 -0500
@@ -1152,6 +1152,7 @@ void MGAStormEngineInit( ScrnInfoPtr pSc
     case PCI_CHIP_MGAG200_EV_PCI:
     case PCI_CHIP_MGAG200_EH_PCI:
     case PCI_CHIP_MGAG200_ER_PCI:	
+    case PCI_CHIP_MGAG200_EH3_PCI:
 	pMga->SrcOrg = 0;
 	OUTREG(MGAREG_SRCORG, pMga->realSrcOrg);
 	OUTREG(MGAREG_DSTORG, pMga->DstOrg);
--- xf86-video-mga-1.6.3/src/mga.h.g200eh3	2016-12-14 05:00:29.918997698 -0500
+++ xf86-video-mga-1.6.3/src/mga.h	2016-12-14 05:05:15.483997776 -0500
@@ -144,6 +144,7 @@ void MGAdbg_outreg32(ScrnInfoPtr, int,in
 #define PCI_CHIP_MGAG200_EH_PCI         0x0533
 #define PCI_CHIP_MGAG200_ER_PCI         0x0534
 #define PCI_CHIP_MGAG200_EW3_PCI        0x0536
+#define PCI_CHIP_MGAG200_EH3_PCI        0x0538
 #define PCI_CHIP_MGAG400                0x0525
 #define PCI_CHIP_MGAG550                0x2527

--- xf86-video-mga-1.6.3/src/mga_g_crtc.c.g200eh3	2016-12-14 05:00:05.862997839 -0500
+++ xf86-video-mga-1.6.3/src/mga_g_crtc.c	2016-12-14 05:14:49.031998593 -0500
@@ -545,7 +545,14 @@ MGAGSetPCLK(xf86CrtcPtr crtc, MgaCrtcSta
                 MGAG200WBComputePLLParam(f_out, &m, &n, &p);
             }
 	} else if (pMga->is_G200EH) {
-	    MGAG200EHComputePLLParam(f_out, &m, &n, &p);
+            if(pMga->Chipset == PCI_CHIP_MGAG200_EH3_PCI)
+            {
+                MGAG200EH3ComputePLLParam(f_out, &m, &n, &p);
+            }
+            else
+            {
+                MGAG200EHComputePLLParam(f_out, &m, &n, &p);
+            }
 	} else if (pMga->is_G200ER) {
 	    MGAG200ERComputePLLParam(f_out, &m, &n, &p);
 	}
@@ -630,6 +637,7 @@ state_set(xf86CrtcPtr crtc, MgaCrtcState
 	break;
 
     case PCI_CHIP_MGAG200_EH_PCI:
+    case PCI_CHIP_MGAG200_EH3_PCI:
 	state->DacRegs[MGA1064_MISC_CTL] =
 	    MGA1064_MISC_CTL_VGA8 |
 	    MGA1064_MISC_CTL_DAC_RAM_CS;
